File: libbambuser-ios-0.9.24-191008/testclient/ViewController.m
/*
 * libbambuser - Bambuser iOS library
 * Copyright 2013 Bambuser AB
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#import "ViewController.h"
#import <AVFoundation/AVFoundation.h>
#import <MobileCoreServices/MobileCoreServices.h>
#import <Photos/Photos.h>

@implementation ViewController

- (id)init {
	if ((self = [super init])) {
		bambuserView = [[BambuserView alloc] initWithPreparePreset: kSessionPresetAuto];
		[bambuserView setDelegate: self];
		bambuserView.applicationId = @"CHANGEME";
		[bambuserView setBroadcastTitle:@"Test broadcast"];

		[bambuserView setTalkback: YES];

		[bambuserView setOrientation: [UIApplication sharedApplication].statusBarOrientation];

		[bambuserView startCapture];

		fileUploader = [[FileUploader alloc] initWithDelegate: self];

		broadcastButton = [UIButton buttonWithType: UIButtonTypeRoundedRect];
		[broadcastButton addTarget:self action:@selector(broadcast) forControlEvents:UIControlEventTouchUpInside];
		[broadcastButton setTitle:@"Broadcast" forState: UIControlStateNormal];

		swapButton = [UIButton buttonWithType: UIButtonTypeRoundedRect];
		[swapButton addTarget: bambuserView action:@selector(swapCamera) forControlEvents:UIControlEventTouchUpInside];
		[swapButton setTitle:@"Swap" forState: UIControlStateNormal];

		talkbackStatus = [[UILabel alloc] init];
		talkbackStatus.textAlignment = NSTextAlignmentLeft;
		talkbackStatus.text = @"";
		talkbackStatus.backgroundColor = [UIColor clearColor];
		talkbackStatus.textColor = [UIColor whiteColor];
		talkbackStatus.shadowColor = [UIColor blackColor];
		talkbackStatus.shadowOffset = CGSizeMake(1, 1);

		uploadButton = [UIButton buttonWithType: UIButtonTypeRoundedRect];
		[uploadButton addTarget:self action:@selector(showImagePicker) forControlEvents:UIControlEventTouchUpInside];
		[uploadButton setTitle:@"Upload" forState: UIControlStateNormal];

		currentViewersLabel = [[UILabel alloc] init];
		currentViewersLabel.textAlignment = NSTextAlignmentLeft;
		currentViewersLabel.text = @"";
		currentViewersLabel.font = [UIFont systemFontOfSize:16];
		currentViewersLabel.backgroundColor = [UIColor clearColor];
		currentViewersLabel.textColor = [UIColor blueColor];

		pinchRecognizer = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(handlePinchGesture:)];
		[pinchRecognizer setDelegate:self];
	}
	return self;
}


- (void) loadView {
	[super loadView];

	[self.view addSubview:bambuserView.view];

	[self.view addSubview:bambuserView.chatView];

	[self.view addSubview:broadcastButton];
	[self.view addSubview:uploadButton];
	[self.view addSubview:currentViewersLabel];
	[self.view addSubview:talkbackStatus];
	if (bambuserView.hasFrontCamera)
		[self.view addSubview:swapButton];
	[self.view addGestureRecognizer: pinchRecognizer];
}

- (void) viewWillLayoutSubviews {
	float statusBarOffset = self.topLayoutGuide.length;

	broadcastButton.frame = CGRectMake(0, 0 + statusBarOffset, 100, 50);
	uploadButton.frame = CGRectMake(self.view.bounds.size.width - 100, 0 + statusBarOffset, 100, 50);
	if (bambuserView.hasFrontCamera)
		swapButton.frame = CGRectMake(0, 50 + statusBarOffset, 100, 50);
	talkbackStatus.frame = CGRectMake(110, 100 + statusBarOffset, 180, 50);
	currentViewersLabel.frame = CGRectMake(self.view.bounds.size.width - 100 , self.view.bounds.size.height - 30, 100, 30);
	[bambuserView setPreviewFrame: CGRectMake(0, 0 + statusBarOffset, self.view.bounds.size.width, self.view.bounds.size.height - statusBarOffset)];
	[bambuserView.chatView setFrame:CGRectMake(0, self.view.bounds.size.height-self.view.bounds.size.height/3, self.view.bounds.size.width, self.view.bounds.size.height/3)];
}

- (UIInterfaceOrientationMask) supportedInterfaceOrientations {
	return UIInterfaceOrientationMaskAllButUpsideDown;
}

- (void) willRotateToInterfaceOrientation:(UIInterfaceOrientation)toInterfaceOrientation duration:(NSTimeInterval)duration {
	[bambuserView setOrientation: toInterfaceOrientation];
}

- (BOOL) shouldAutorotate {
	return broadcastButton.enabled && ![broadcastButton.currentTitle isEqualToString: @"Stop"];
}

- (BOOL) prefersStatusBarHidden {
	return NO;
}

- (void)handlePinchGesture: (UIPinchGestureRecognizer *)sender {
	if (sender.state == UIGestureRecognizerStateBegan) {
		initialZoom = bambuserView.zoom;
	}
	bambuserView.zoom = initialZoom * sender.scale;
}

-(void) showImagePicker {
	UIImagePickerController *picker = [[UIImagePickerController alloc] init];
	picker.delegate = self;
	picker.allowsEditing = NO;
	picker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
	picker.mediaTypes = [[NSArray alloc] initWithObjects: (NSString *) kUTTypeImage, (NSString*) kUTTypeMovie, nil];
	picker.videoQuality = UIImagePickerControllerQualityTypeIFrame1280x720;
	if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
		// This needs to be stored in a member, since it gets deallocated (while still visible)
		// if nothing keeps a reference to it.
		popover = [[UIPopoverController alloc] initWithContentViewController:picker];
		[popover presentPopoverFromRect:uploadButton.frame inView:self.view permittedArrowDirections:UIPopoverArrowDirectionAny animated:YES];
	} else {
		[self presentViewController:picker animated:YES completion:NULL];
	}
}

- (void) imagePickerController: (UIImagePickerController *) picker didFinishPickingMediaWithInfo: (NSDictionary *) info {
	NSURL *referenceURL = [info objectForKey:UIImagePickerControllerReferenceURL];

	PHFetchResult *fetchResult = [PHAsset fetchAssetsWithALAssetURLs:@[referenceURL] options:nil];
	if (fetchResult.firstObject) {
		[self uploadAsset:fetchResult.firstObject mediaInfo:info];
	} else {
		NSLog(@"Asset not found");
	}
	[picker dismissViewControllerAnimated:YES completion:NULL];
}

- (void) imagePickerControllerDidCancel: (UIImagePickerController *) picker {
	[picker dismissViewControllerAnimated:YES completion:NULL];
}

- (void) uploadAsset: (NSObject*) asset mediaInfo: (NSDictionary*)info {

	// Custom key-values to associate with your upload.
	NSDictionary *dictionary = @{ @"exampleKey": @"exampleValue"};

	NSMutableDictionary *ticketAssets = [NSMutableDictionary dictionary];
	[ticketAssets addEntriesFromDictionary: @{
		@"info" : info,
		@"fileType" : ([[info valueForKey: UIImagePickerControllerMediaType] isEqualToString:@"public.image"] ? @"image" : @"video"),
		@"author" : @"JOHN CHANGEME",
		@"title" : @"Test broadcast",
		@"deleteAfterUpload" : [NSNumber numberWithBool: YES],
		@"dictionary" : dictionary,
		@"asset" : asset
	}];
	if (bambuserView.applicationId)
		[ticketAssets setValue: bambuserView.applicationId forKey: @"applicationId"];
	[fileUploader performSelectorInBackground: @selector(getTicketAndUpload:) withObject: ticketAssets];
}

- (void) broadcast {
	[broadcastButton setTitle:@"Connecting" forState:UIControlStateNormal];
	[broadcastButton removeTarget:nil action:NULL forControlEvents:UIControlEventTouchUpInside];
	[broadcastButton addTarget:bambuserView action:@selector(stopBroadcasting) forControlEvents:UIControlEventTouchUpInside];
	[bambuserView startBroadcasting];
}

- (void) broadcastButtonStop {
	[broadcastButton setTitle:@"Stop" forState:UIControlStateNormal];
	[broadcastButton removeTarget:nil action:NULL forControlEvents:UIControlEventTouchUpInside];
	[broadcastButton addTarget:bambuserView action:@selector(stopBroadcasting) forControlEvents:UIControlEventTouchUpInside];
}

- (void) broadcastStarted {
	[self performSelectorOnMainThread:@selector(broadcastButtonStop) withObject:nil waitUntilDone:YES];
	NSLog(@"Received broadcastStarted signal");
}

- (void) broadcastButtonBroadcast {
	[broadcastButton setTitle:@"Broadcast" forState:UIControlStateNormal];
	[broadcastButton removeTarget:nil action:NULL forControlEvents:UIControlEventTouchUpInside];
	[broadcastButton addTarget:self action:@selector(broadcast) forControlEvents:UIControlEventTouchUpInside];
}

- (void) broadcastStopped {
	[self performSelectorOnMainThread:@selector(broadcastButtonBroadcast) withObject:nil waitUntilDone:YES];
	currentViewersLabel.text = @"";
	NSLog(@"Received broadcastStopped signal");
}

- (void) recordingComplete: (NSString*) filename {
	NSLog(@"Got filename: %@", filename);
	NSURL *url = [[NSURL alloc] initFileURLWithPath:filename];

	/*
	 * We need to do this check since the performChanges call returns
	 * before the user has granted access in cases it is not yet determined
	 */
	PHAuthorizationStatus authorizationStatus = [PHPhotoLibrary authorizationStatus];
	switch (authorizationStatus) {
		case PHAuthorizationStatusNotDetermined: {
			[PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
				if (status == PHAuthorizationStatusAuthorized) {
					[self recordingComplete:filename];
				}
			}];

			break;
		}
		case PHAuthorizationStatusAuthorized: {
			[[PHPhotoLibrary sharedPhotoLibrary] performChanges:^ {
				[PHAssetChangeRequest creationRequestForAssetFromVideoAtFileURL:url];
			} completionHandler:^(BOOL success, NSError *error) {
				[self saveComplete:success fileName:filename];
			}];
			break;
		}
		case PHAuthorizationStatusRestricted:
		case PHAuthorizationStatusDenied: {
			/*
			 *  This scenario is likely to happen if the user denies access to the camera roll.
			 *  If exporting to camera roll is your only option, make sure to recover or remove this video file.
			 */
			[self showError:@"Video not saved locally, access to Photos not granted"];
			break;
		}
		default:
			break;
	}
}

- (void) saveComplete: (BOOL)success fileName:(NSString*) filename {
	if (success) {
		NSLog(@"Exported recording from %@ to camera roll", filename);
		NSFileManager* fileManager = [NSFileManager defaultManager];
		if ([fileManager removeItemAtPath:filename error:nil]) {
			NSLog(@"Removing file from %@", filename);
		}
	} else {
		/*
		 *  This scenario is likely to happen if the user denies access to the camera roll.
		 *  If exporting to camera roll is your only option, make sure to recover or remove this video file.
		 */
		NSLog(@"Failed to export recording from %@", filename);
	}
}

- (void) showError: (NSString*) errorMessage {
	UIAlertController *alertController = [UIAlertController alertControllerWithTitle: @"Error" message: errorMessage preferredStyle: UIAlertControllerStyleAlert];
	UIAlertAction *okAction = [UIAlertAction actionWithTitle: @"OK" style: UIAlertActionStyleDefault handler: nil];
	[alertController addAction: okAction];
	[self presentViewController: alertController animated: YES completion: nil];
}

- (void) bambuserError: (enum BambuserError)errorCode message:(NSString*)errorMessage {
	switch (errorCode) {
		case kBambuserErrorServerFull:
		case kBambuserErrorIncorrectCredentials:
		case kBambuserErrorConnectionLost:
		case kBambuserErrorUnableToConnect:
		case kBambuserErrorBroadcastTicketFailed:
			// Enable broadcastbutton on connection error
			[self performSelectorOnMainThread:@selector(broadcastButtonBroadcast) withObject:nil waitUntilDone:YES];
			break;
		case kBambuserErrorServerDisconnected:
		case kBambuserErrorLocationDisabled:
		case kBambuserErrorNoCamera:
		default:
			break;
	}
	[self performSelectorOnMainThread:@selector(showError:) withObject:errorMessage waitUntilDone:NO];
}

- (void) chatMessageReceived:(NSString *)message {
	[bambuserView displayMessage: message];
}

- (void) talkbackRequest: (NSString*) request caller: (NSString*) caller talkbackID: (int) talkbackID {
	UIAlertController *talkbackRequest = [UIAlertController alertControllerWithTitle: caller message: request preferredStyle :UIAlertControllerStyleAlert];
	UIAlertAction *acceptAlert = [UIAlertAction actionWithTitle: @"Accept" style: UIAlertActionStyleDefault handler: ^(UIAlertAction * _Nonnull action) {
		[self->bambuserView acceptTalkbackRequest: (int) talkbackID];
	}];
	UIAlertAction *declineAlert = [UIAlertAction actionWithTitle: @"Decline" style: UIAlertActionStyleDestructive handler: ^(UIAlertAction * _Nonnull action) {
		[self->bambuserView declineTalkbackRequest: (int) talkbackID];
	}];
	[talkbackRequest addAction: acceptAlert];
	[talkbackRequest addAction: declineAlert];

	[self presentViewController: talkbackRequest animated: YES completion: nil];
}

- (void) talkbackStateChanged: (enum TalkbackState)state {
	// Update label showing talkback state
	switch (state) {
		case kTalkbackNeedsAccept:
			talkbackStatus.text = @"Talkback pending";
			break;
		case kTalkbackAccepted:
			talkbackStatus.text = @"Talkback accepted";
			break;
		case kTalkbackPlaying:
			talkbackStatus.text = @"Talkback playing";
			break;
		default:
			talkbackStatus.text = @"";
	}

	// Add or remove button for ending talkback
	if (state == kTalkbackPlaying) {
		endTalkbackButton = [UIButton buttonWithType: UIButtonTypeRoundedRect];
		[endTalkbackButton addTarget: bambuserView action:@selector(endTalkback) forControlEvents:UIControlEventTouchUpInside];
		float statusBarOffset = self.topLayoutGuide.length;
		endTalkbackButton.frame = CGRectMake(0, 150 + statusBarOffset, 100, 50);
		[endTalkbackButton setTitle:@"End talkback" forState: UIControlStateNormal];
		[self.view addSubview:endTalkbackButton];
	} else if (endTalkbackButton != nil) {
		[endTalkbackButton removeFromSuperview];
		endTalkbackButton = nil;
	}
}

- (void) currentViewerCountUpdated: (int) viewers {
	currentViewersLabel.text = [NSString stringWithFormat: @"Viewers: %d", viewers];
}

- (void) totalViewerCountUpdated: (int) viewers {
	NSLog(@"Total viewers: %d", viewers);
}

-(void) uploadStarted: (NSNumber*) async {
	uploadAlertController = [UIAlertController alertControllerWithTitle: @"Uploading" message: @"" preferredStyle:UIAlertControllerStyleAlert];

	if ([async boolValue]) {
		UIAlertAction *cancel = [UIAlertAction actionWithTitle: @"Cancel" style: UIAlertActionStyleCancel handler: ^(UIAlertAction * _Nonnull action) {
			[self uploadEnded];
			[self->fileUploader cancelUpload];
		}];
		[uploadAlertController addAction: cancel];
	}
	progressBar = [[UIProgressView alloc] initWithFrame:CGRectMake(25.0f, 60.0f, 230.0f, 90.0f)];
	[progressBar setProgressViewStyle: UIProgressViewStyleBar];
	[uploadAlertController.view addSubview: progressBar];
	[self presentViewController: uploadAlertController animated:YES completion: nil];
}

-(void) uploadUpdated: (NSNumber*) progress {
	[progressBar setProgress: [progress floatValue]];
}

-(void) uploadEnded {
	[uploadAlertController dismissViewControllerAnimated: YES completion: nil];
	uploadAlertController = nil;
	progressBar = nil;
}

-(void) uploadFailed {
	[uploadAlertController dismissViewControllerAnimated: YES completion: nil];
	uploadAlertController = nil;
	progressBar = nil;
	[self performSelectorOnMainThread:@selector(showError:) withObject:@"Upload failed" waitUntilDone:NO];
}

@end

File: libbambuser-ios-0.9.24-191008/testclient-swift/ViewController.swift
/*
 * libbambuser - Bambuser iOS library
 * Copyright 2015 Bambuser AB
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

import UIKit
import MobileCoreServices
import Photos

class ViewController: UIViewController, BambuserViewDelegate, FileUploaderDelegate, UINavigationControllerDelegate, UIImagePickerControllerDelegate, UIGestureRecognizerDelegate {
	var bambuserView: BambuserView
	var broadcastButton: UIButton
	var swapButton: UIButton
	var uploadButton: UIButton
	var endTalkbackButton: UIButton
	var talkbackStatus: UILabel
	var currentViewersLabel: UILabel
	var pinchRecognizer: UIPinchGestureRecognizer
	var initialZoom: Float
	var popover: UIPopoverController?
	var fileUploader: FileUploader?
	var uploadAlertController: UIAlertController?
	var progressBar: UIProgressView?

	required init?(coder aDecoder: NSCoder) {
		bambuserView = BambuserView(preparePreset: kSessionPresetAuto)
		broadcastButton = UIButton(type: UIButtonType.system)
		swapButton = UIButton(type: UIButtonType.system)
		uploadButton = UIButton(type: UIButtonType.system)
		endTalkbackButton = UIButton(type: UIButtonType.system)
		talkbackStatus = UILabel()
		currentViewersLabel = UILabel()
		pinchRecognizer = UIPinchGestureRecognizer()
		initialZoom = 0.0

		super.init(coder: aDecoder)

		bambuserView.delegate = self;
		bambuserView.applicationId = "CHANGEME"
		bambuserView.broadcastTitle = "Test broadcast"

		bambuserView.talkback = true

		bambuserView.orientation = UIApplication.shared.statusBarOrientation

		bambuserView.startCapture()

		fileUploader = FileUploader(_delegate: self)

		broadcastButton.addTarget(self, action: #selector(ViewController.broadcast), for: UIControlEvents.touchUpInside)
		broadcastButton.setTitle("Broadcast", for: UIControlState())

		swapButton.addTarget(bambuserView, action: #selector(BambuserView.swapCamera), for: UIControlEvents.touchUpInside)
		swapButton.setTitle("Swap", for: UIControlState())

		talkbackStatus.textAlignment = NSTextAlignment.left;
		talkbackStatus.text = ""
		talkbackStatus.backgroundColor = UIColor.clear
		talkbackStatus.textColor = UIColor.white
		talkbackStatus.shadowColor = UIColor.black
		talkbackStatus.shadowOffset = CGSize(width: 1, height: 1)

		uploadButton.addTarget(self, action: #selector(ViewController.showImagePicker), for: UIControlEvents.touchUpInside)
		uploadButton.setTitle("Upload", for: UIControlState())
		pinchRecognizer = UIPinchGestureRecognizer(target: self, action: #selector(ViewController.handlePinchGesture(_:)))

		currentViewersLabel.textAlignment = NSTextAlignment.left
		currentViewersLabel.text = ""
		currentViewersLabel.font = UIFont.systemFont(ofSize: 16)
		currentViewersLabel.backgroundColor = UIColor.clear
		currentViewersLabel.textColor = UIColor.blue
	}

	override func loadView() {
		super.loadView()
		self.view.addSubview(bambuserView.view)
		self.view.addSubview(bambuserView.chatView)

		self.view.addSubview(broadcastButton)
		self.view.addSubview(uploadButton)
		self.view.addSubview(talkbackStatus)
		self.view.addSubview(currentViewersLabel)
		if (bambuserView.hasFrontCamera) {
			self.view.addSubview(swapButton)
		}
		self.view.addGestureRecognizer(pinchRecognizer);
	}

	override func viewDidLoad() {
		super.viewDidLoad()
		// Do any additional setup after loading the view, typically from a nib.
	}

	override func viewWillLayoutSubviews() {
		var statusBarOffset : CGFloat = 0.0
		statusBarOffset = CGFloat(self.topLayoutGuide.length)

		broadcastButton.frame = CGRect(x: 0.0, y: 0.0 + statusBarOffset, width: 100.0, height: 50.0);
		uploadButton.frame = CGRect(x: self.view.bounds.size.width - 100.0, y: 0.0 + statusBarOffset, width: 100.0, height: 50.0);
		if (bambuserView.hasFrontCamera) {
			swapButton.frame = CGRect(x: 0.0, y: 50.0 + statusBarOffset, width: 100.0, height: 50.0);
		}
		talkbackStatus.frame = CGRect(x: 110.0, y: 100.0 + statusBarOffset, width: 180.0, height: 50.0);
		currentViewersLabel.frame = CGRect(x: self.view.bounds.size.width - 100 , y: self.view.bounds.size.height - 30, width: 100, height: 30)
		bambuserView.previewFrame = CGRect(x: 0.0, y: 0.0 + statusBarOffset, width: self.view.bounds.size.width, height: self.view.bounds.size.height - statusBarOffset)
		bambuserView.chatView.frame = CGRect(x: 0.0, y: self.view.bounds.size.height-self.view.bounds.size.height/3.0, width: self.view.bounds.size.width, height: self.view.bounds.size.height/3.0)
	}

	override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
		return UIInterfaceOrientationMask.allButUpsideDown
	}

	override func willRotate(to toInterfaceOrientation: UIInterfaceOrientation, duration: TimeInterval) {
		bambuserView.setOrientation(toInterfaceOrientation, previewOrientation: toInterfaceOrientation)
	}

	override var shouldAutorotate : Bool {
		return broadcastButton.isEnabled && broadcastButton.currentTitle != "Stop";
	}

	override var prefersStatusBarHidden : Bool {
		return false
	}

	override func didReceiveMemoryWarning() {
		super.didReceiveMemoryWarning()
		// Dispose of any resources that can be recreated.
	}

	@objc func showImagePicker() {
		let picker = UIImagePickerController()
		picker.delegate = self
		picker.allowsEditing = false
		picker.sourceType = UIImagePickerControllerSourceType.photoLibrary
		picker.mediaTypes = [kUTTypeImage as String, kUTTypeMovie as String]
		picker.videoQuality = UIImagePickerControllerQualityType.typeIFrame1280x720
		if (UIDevice.current.userInterfaceIdiom == UIUserInterfaceIdiom.pad) {
			popover = UIPopoverController(contentViewController: picker)
			popover?.present(from: uploadButton.frame, in: self.view, permittedArrowDirections: UIPopoverArrowDirection.any, animated: true)
		} else {
			self.present(picker, animated: true, completion: nil)
		}
	}

	func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {

		let referenceURL = info[UIImagePickerControllerReferenceURL] as! URL

		let fetchResult:PHFetchResult = PHAsset.fetchAssets(withALAssetURLs: [referenceURL], options: nil)
		if (fetchResult.firstObject != nil) {
			self.uploadAsset(fetchResult.firstObject!, info: info as [String : AnyObject])
		} else {
			print("Asset not found")
		}

		picker.dismiss(animated: true, completion: nil)
	}

	func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
		picker.dismiss(animated: true, completion: nil);
	}

	func uploadAsset(_ asset: NSObject, info:[String : AnyObject]) {
		let dictionary = ["exampleKey" : "exampleValue"]
		var ticketAssets = [
			"info" : info,
			"fileType" : (info[UIImagePickerControllerMediaType] as! String == "public.image" ? "image" : "video"),
			"author" : "JOHN CHANGEME",
			"title" : "Test broadcast",
			"deleteAfterUpload" : true,
			"dictionary" : dictionary,
			"asset" : asset
		] as [String : Any]

		if (!(self.bambuserView.applicationId ?? "").isEmpty) {
			ticketAssets["applicationId"] = self.bambuserView.applicationId
		}
		DispatchQueue.main.async {
			self.fileUploader!.getTicketAndUpload(ticketAssets as Dictionary<String, AnyObject>)
		}
	}

	@objc func broadcast() {
		broadcastButton.setTitle("Connecting", for: UIControlState())
		broadcastButton.isEnabled = false
		bambuserView.startBroadcasting()
	}

	func broadcastButtonStop() {
		broadcastButton.setTitle("Stop", for: UIControlState())
		broadcastButton.isEnabled = true
		broadcastButton.removeTarget(nil, action: nil, for: UIControlEvents.touchUpInside)
		broadcastButton.addTarget(bambuserView, action: #selector(BambuserView.stopBroadcasting), for: UIControlEvents.touchUpInside)
	}

	func broadcastStarted() {
		self.broadcastButtonStop()
		print("Received broadcastStarted signal")
	}

	func broadcastButtonBroadcast() {
		broadcastButton.setTitle("Broadcast", for: UIControlState())
		broadcastButton.isEnabled = true
		broadcastButton.removeTarget(nil, action: nil, for: UIControlEvents.touchUpInside)
		broadcastButton.addTarget(self, action: #selector(ViewController.broadcast), for: UIControlEvents.touchUpInside)
	}

	func broadcastStopped() {
		self.broadcastButtonBroadcast()
		self.currentViewersLabel.text = ""
	}

	func recordingComplete(_ filename : String) {
		let url = URL(fileURLWithPath: filename)
		/*
		 * We need to do this check since the performChanges call returns
		 * before the user has granted access in cases it is not yet determined
		 */
		let authorizationStatus = PHPhotoLibrary.authorizationStatus()
		switch (authorizationStatus) {
		case PHAuthorizationStatus.notDetermined:
			PHPhotoLibrary.requestAuthorization({ (status) in
				if (status == PHAuthorizationStatus.authorized) {
					self.recordingComplete(filename)
				}
			})
			break
		case PHAuthorizationStatus.authorized:
			PHPhotoLibrary.shared().performChanges({
				PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: url)
				}, completionHandler: { (success, error) in
					self.saveComplete(success, filename: filename)
			})
			break
		case PHAuthorizationStatus.restricted:
			fallthrough
		case PHAuthorizationStatus.denied:
			fallthrough
		default:
			/*
			 *  This scenario is likely to happen if the user denies access to the camera roll.
			 *  If exporting to camera roll is your only option, make sure to recover or remove this video file.
			 */
			self.showError("Video not saved locally, access to Photos not granted")
			break
		}
	}

	func saveComplete(_ success:Bool, filename:String) {
		if (success) {
			print("Exported recording from " + filename + " to camera roll")
			let fileManager = FileManager.default
			do {
				try fileManager.removeItem(atPath: filename)
				print("Removing file from " + filename)
			} catch {
				print("Failed to remove temporary recording")
			}
		} else {
			/*
			 *  This scenario is likely to happen if the user denies access to the camera roll.
			 *  If exporting to camera roll is your only option, make sure to recover or remove this video file.
			 */
			print("Failed to export recording from " + filename)
		}
	}

	func showError(_ errorMessage : String) {
		let alertController:UIAlertController =  UIAlertController(title: "Error", message: errorMessage, preferredStyle: .alert)
		let cancelAction = UIAlertAction(title: "OK", style: .default, handler: nil)
		alertController.addAction(cancelAction)
		self.present(alertController, animated: true, completion: nil)
	}

	func bambuserError(_ code: BambuserError, message: String) {
		switch (code) {
			case kBambuserErrorServerFull,
			kBambuserErrorIncorrectCredentials,
			kBambuserErrorConnectionLost,
			kBambuserErrorUnableToConnect:
				// Enable broadcastbutton on connection error
				DispatchQueue.main.async {
					self.broadcastButtonBroadcast()
				}
			break
		case kBambuserErrorServerDisconnected, kBambuserErrorLocationDisabled, kBambuserErrorNoCamera:
			break;
		default:
			break
		}
		DispatchQueue.main.async {
			self.showError(message as String)
		}
	}

	func chatMessageReceived(_ message : String) {
		bambuserView.displayMessage(String(message))
	}

	func talkbackRequest(_ request : String, caller : String, talkbackID : Int32) {
		let talkbackRequest = UIAlertController(title: caller, message: request, preferredStyle: .alert)
		let acceptAction = UIAlertAction(title: "Accept", style: .default) { (action:UIAlertAction) in
			self.bambuserView.acceptTalkbackRequest(talkbackID)
		}
		let declineAction = UIAlertAction(title: "Decline", style: .destructive) { (action:UIAlertAction) in
			self.bambuserView.declineTalkbackRequest(talkbackID)
		}
		talkbackRequest.addAction(acceptAction)
		talkbackRequest.addAction(declineAction)
		self.present(talkbackRequest, animated: true, completion: nil)
	}

	func talkbackStateChanged(_ state: TalkbackState) {
		switch (state) {
		case kTalkbackNeedsAccept:
			talkbackStatus.text = "Talkback pending";
			break;
		case kTalkbackAccepted:
			talkbackStatus.text = "Talkback accepted";
			break;
		case kTalkbackPlaying:
			talkbackStatus.text = "Talkback playing";
			break;
		default:
			talkbackStatus.text = "";
		}

		// Add or remove button for ending talkback
		if (state == kTalkbackPlaying) {
			endTalkbackButton = UIButton(type: UIButtonType.system)
			endTalkbackButton.addTarget(bambuserView, action: #selector(BambuserView.endTalkback), for: UIControlEvents.touchUpInside)
			let statusBarOffset = self.topLayoutGuide.length;
			endTalkbackButton.frame = CGRect(x: 0.0, y: 150.0 + statusBarOffset, width: 100.0, height: 50.0);
			endTalkbackButton.setTitle("End talkback", for: UIControlState())
			self.view.addSubview(endTalkbackButton)
		} else {
			endTalkbackButton.removeFromSuperview()
		}
	}

	func currentViewerCountUpdated(_ viewers: Int32) {
		currentViewersLabel.text = "Viewers: \(viewers)"
	}

	func totalViewerCountUpdated(_ viewers: Int32) {
		print("Total viewers: \(viewers)")
	}

	func uploadStarted(_ async: NSNumber) {
		uploadAlertController =  UIAlertController(title: "Uploading", message: "", preferredStyle: .alert)
		if (async.boolValue) {
			let cancelAction = UIAlertAction(title: "Cancel", style: .cancel) { (action:UIAlertAction) in
				self.uploadEnded()
				self.fileUploader?.cancelUpload()
			}
			uploadAlertController!.addAction(cancelAction)
		}
		progressBar = UIProgressView(frame: CGRect(x: 25.0, y: 50.0, width: 230.0, height: 90.0))
		progressBar?.progressViewStyle = UIProgressViewStyle.bar
		uploadAlertController?.view.addSubview(progressBar!)
		self.present(uploadAlertController!, animated: true, completion: nil)
	}

	func uploadUpdated(_ progress: NSNumber) {
		progressBar?.setProgress(progress.floatValue, animated: true)
	}

	func uploadEnded() {
		uploadAlertController?.dismiss(animated: true, completion: nil)
		uploadAlertController = nil;
		progressBar = nil;
	}

	func uploadFailed() {
		uploadAlertController?.dismiss(animated: true, completion: nil)
		uploadAlertController = nil;
		progressBar = nil;
		DispatchQueue.main.async {
			self.showError("Upload failed")
		}
	}

	@objc func handlePinchGesture(_ sender : UIPinchGestureRecognizer) {
		if (sender.state == UIGestureRecognizerState.began) {
			initialZoom = bambuserView.zoom
		}
		bambuserView.zoom = initialZoom * Float(sender.scale)
	}
}

